### PR TITLE
resources: Customizable contexts.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
This wraps the resource acquisition execution context in a type, for a few reasons:

1. It makes it easier to ensure the execution context does not leak into the main part of the application.
2. We can more easily identify when a strange execution context is used, such as the `DirectExecutionContext`.
3. We can potentially add extra information to this type, such as the logging context.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by SamirTalwar-DA** [comment]
BTW, I tried to make it wrap a phantom-typed execution context (i.e. `ExecutionContext[Acquisition]`) to make it even harder to leak, but I had to change so much to make it work and it was an uphill struggle due to IntelliJ not being able to recognize "no such method: `flatMap`" vs. the types not lining up perfectly to use the relevant implicit.

TLDR implicits make everything fun.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala [hunk 1]
```diff
@@ -211,7 +207,7 @@ object JdbcIndexer {
     }
 
   private def contextFor(offset: Offset, update: Update): Map[String, String] =
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala:209]: Please rename this as well.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala [hunk 2]
```diff
@@ -120,9 +116,9 @@ object JdbcIndexer {
       ledgerDao.initializeLedger(providedLedgerId)
     }
 
-    private def initOrCheckParticipantId(dao: LedgerDao)(
-        implicit executionContext: ExecutionContext,
-    ): Future[Unit] = {
+    private def initOrCheckParticipantId(
+        dao: LedgerDao,
+    )(implicit resourceContext: ResourceContext): Future[Unit] = {
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala:121]: There's inconsistency in how you name your implicit resource contexts. I would recommend sticking to the naming you used here (`resourceContext`), since we also have other contexts we pass around.

File: ledger/participant-integration-api/src/main/scala/platform/store/FlywayMigrations.scala [hunk 3]
```diff
@@ -3,27 +3,27 @@
...
 private[platform] class FlywayMigrations(jdbcUrl: String)(implicit loggingContext: LoggingContext) {
   private val logger = ContextualizedLogger.get(this.getClass)
 
   private val dbType = DbType.jdbcType(jdbcUrl)
 
-  def validate()(implicit executionContext: ExecutionContext): Future[Unit] =
+  def validate()(implicit resourceContext: ResourceContext): Future[Unit] =
     dataSource.use { ds =>
-      Future {
+      Future.successful {
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/FlywayMigrations.scala:26]: Can you double check that this change is ok? I believe it's ok because I guess `dataSource.use` wraps the `Future` so that the `Throwable` is caught correctly, but it's difficult to say from GitHub. Same for the two following analogous changes.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala [hunk 4]
```diff
@@ -55,6 +54,7 @@ private[dao] trait JdbcLedgerDaoBackend extends AkkaBeforeAndAfterAll { this: Su
   override protected def beforeAll(): Unit = {
     super.beforeAll()
     implicit val executionContext: ExecutionContext = system.dispatcher
+    implicit val resourceContext: ResourceContext = ResourceContext(executionContext)
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala:57]: The double implicit execution context here seems like asking for trouble. Since this PR is already giant it would probably make sense to use the `ResourceContext` in `AkkaBeforeAndAfterAll`. You can tackle it in a second PR if you think it's a better idea, but it would probably make sense to make the change ASAP.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/GrpcServerSpec.scala [hunk 5]
```diff
@@ -11,16 +11,18 @@ import com.daml.ledger.client.configuration.{
...
 import com.daml.platform.apiserver.GrpcServerSpec._
 import com.daml.platform.hello.{HelloRequest, HelloServiceGrpc}
 import com.daml.ports.Port
-import com.daml.resources.ResourceOwner
 import com.google.protobuf.ByteString
 import io.grpc.ManagedChannel
 import org.scalatest.{AsyncWordSpec, Matchers}
 
 final class GrpcServerSpec extends AsyncWordSpec with Matchers {
+  private implicit val resourceContext: ResourceContext = ResourceContext(executionContext)
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/GrpcServerSpec.scala:24]: Should we have our own `AsyncWordSpec` that uses the `ResourceContext`? Having two implicit contexts in the mix seems like a recipe for trouble.

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/AbstractResourceOwner.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+/**
+  * A ResourceOwner of type [[A]] is can acquire a [[Resource]] of the same type and its operations
+  * are applied to the [[Resource]] after it has been acquired.
+  *
+  * @tparam A The [[Resource]] value type.
+  */
+@FunctionalInterface
```
Comment (stefanobaghino-da) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/AbstractResourceOwner.scala:15]: Are you doing this to have some kind of interoperability with Java? With `acquire` using an implicit parameter I'm not entirely sure of the usefulness of this.

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/HasExecutionContext.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+import scala.concurrent.ExecutionContext
+
+trait HasExecutionContext[Context] {
+  def executionContext(context: Context): ExecutionContext
+}
+
+object HasExecutionContext {
+  def apply[Context: HasExecutionContext]: HasExecutionContext[Context] =
+    Option(implicitly[HasExecutionContext[Context]]).get
```
Comment (stefanobaghino-da) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/HasExecutionContext.scala:14]: `implicitly` looks neat, but I would make the evidence explicit and just use that.

**Review by stefanobaghino-da — COMMENTED** [review]
I would approve, but this PR is too large in scope to be merged with _at least_ a second review.

**Comment by stefanobaghino-da** [comment]
> BTW, I tried to make it wrap a phantom-typed execution context (i.e. `ExecutionContext[Acquisition]`) to make it even harder to leak, but I had to change so much to make it work and it was an uphill struggle due to IntelliJ not being able to recognize "no such method: `flatMap`" vs. the types not lining up perfectly to use the relevant implicit.
> 
> TLDR implicits make everything fun.

I'm a bit worried this is getting a bit too clever.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala:121]: I tried to keep it as `resourceContext` except when it's the implicit parameter of a `ResourceOwner#acquire`, in which case it's `context`. Is this too confusing?

File: ledger/participant-integration-api/src/main/scala/platform/store/FlywayMigrations.scala (hunk 3 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/FlywayMigrations.scala:26]: It is, because `use` calls the behavior inside a `flatMap`. There's a unit test for it.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala:57]: That's a good idea. I'll make that change.

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/AbstractResourceOwner.scala (hunk 6 above)
Comment (SamirTalwar-DA) [reply libs-scala/resources/src/main/scala/com/digitalasset/resources/AbstractResourceOwner.scala:15]: Yeah, this probably isn't true anyway. I shall delete it. (It was there before.)

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/GrpcServerSpec.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/GrpcServerSpec.scala:24]: I could add a helper trait: `with SpecResourceContext` or something.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoBackend.scala:57]: Decided against this and went with the helper trait for these too, because otherwise `AkkaBeforeAndAfterAll` starts to depend on stuff in `//ledger`.

**Comment by SamirTalwar-DA** [comment]
> I'm a bit worried this is getting a bit too clever.

I came to the same conclusion and threw away that particular changeset.

**Review by stefanobaghino-da — APPROVED** [review]

**Comment by stefanobaghino-da** [comment]
> /azp run digital-asset.daml (macOS)

You wish! 😂

**Comment by SamirTalwar-DA** [comment]
Yeah, the answer to [this question](https://developercommunity.visualstudio.com/content/problem/1225480/failed-jobs-cannot-be-rerun-via-the-github-checks.html) made me think we could get away with this.
